python/ml4ir/base/model/optimizer.py

```diff
@@ -5,58 +5,61 @@
 
     Arguments:
-        optimizer_key: string optimizer name to be used as defined under ml4ir.base.config.keys.OptimizerKey
-        learning_rate: floating point learning rate for the optimizer
-        learning_rate_decay: floating point rate at which the learning rate will be decayed every learning_rate_decay_steps
-        learning_rate_decay_steps: int representing number of iterations after which learning rate will be decreased exponentially
-        gradient_clip_value: float value representing the clipvalue for gradient updates. Not setting this to a reasonable value based on the model will lead to gradient explosion and NaN losses.
+        optimizer_key: string optimizer name to be used as defined
+         under tensorflow.keras.optimizers
+        with_exp_decay: Set to True, to add Exponential Decay schedule
+        learning_rate_decay,learning_rate_decay, learning_rate_decay_steps:
+        Params controlling the decay schedule,
+         cf. https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/ExponentialDecay
+
+    You may pass any other valid arguments to the function to control aspects of the optimizer,
+    according to the official tensforflow documentation.
 
     References:
         https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer
         https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/ExponentialDecay
 
+
+    Examples:
+        get_optimizer() -> SGD optimizer
+        get_optimizer('Adam', learning_rate=0.002) -> Adam, with lr=0.002
     FIXME:
         Define all arguments overriding tensorflow defaults in a separate file
         for visibility with ml4ir users
+
     """
     # Define an exponential learning rate decay schedule
     learning_rate_schedule = ExponentialDecay(
-        learning_rate,
+        initial_learning_rate,
         decay_steps=learning_rate_decay_steps,
         decay_rate=learning_rate_decay,
         staircase=True,
     )
 
-    if optimizer_key == OptimizerKey.ADAM:
-        return tf_optimizers.Adam(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.NADAM:
-        return tf_optimizers.Nadam(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.ADAGRAD:
-        return tf_optimizers.Adagrad(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.SGD:
-        return tf_optimizers.SGD(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    elif optimizer_key == OptimizerKey.RMS_PROP:
-        return tf_optimizers.RMSprop(
-            learning_rate=learning_rate_schedule, clipvalue=gradient_clip_value
-        )
-    else:
-        raise ValueError("illegal Optimizer key: " + optimizer_key)
+    try:
```

Comment:
This gist of the PR is here. 
These lines of code enable you to use any tf optimizer and pass any keyword arguments as long as you respect the tf names. So to use tensorflow.keras.optimizers.Adam, the key needs to be 'Adam' ('adam') will fail. Same goes for any specific keyword args of the optimizer.

Reply:
If you like this, we should do similarly with metrics, layers etc. 
Some pointers where we could improve a lot: 
- https://github.com/salesforce/ml4ir/blob/master/python/ml4ir/base/model/architectures/dnn.py#L48
- https://github.com/salesforce/ml4ir/blob/master/python/ml4ir/applications/ranking/model/metrics/metric_factory.py